hashes/src/sha256.rs

```diff
@@ -133,33 +152,61 @@ impl Hash {
     /// Computes hash from `bytes` in `const` context.
     ///
     /// Warning: this function is inefficient. It should be only used in `const` context.
-    pub const fn const_hash(bytes: &[u8]) -> Self { Hash(Midstate::const_hash(bytes, true).0) }
+    pub const fn const_hash(bytes: &[u8]) -> Self { Hash(Midstate::const_hash(bytes, true).bytes) }
 }
 
 /// Output of the SHA256 hash function.
-#[derive(Copy, Clone, PartialEq, Eq, Default, PartialOrd, Ord, Hash)]
-pub struct Midstate(pub [u8; 32]);
+///
+/// This type should not be used directly unless you really know what you're doing i.e., you
+/// probably should be a professional cryptographer.
+#[derive(Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Midstate {
+    /// Raw bytes of the midstate i.e., the already-hashed contents of the hash engine.
+    bytes: [u8; 32],
+    /// Number of bytes hashed to achieve this midstate.
+    // INVARIANT must always be a multiple of 64.
+    length: usize,
+}
 
 impl Midstate {
-    /// Length of the midstate, in bytes.
-    const LEN: usize = 32;
-
     /// Construct a new [`Midstate`] from the inner value.
-    pub const fn from_byte_array(inner: [u8; 32]) -> Self { Midstate(inner) }
-
-    /// Copies a byte slice into the [`Midstate`] object.
-    pub fn from_slice(sl: &[u8]) -> Result<Midstate, FromSliceError> {
-        if sl.len() != Self::LEN {
-            Err(FromSliceError { expected: Self::LEN, got: sl.len() })
-        } else {
-            let mut ret = [0; 32];
-            ret.copy_from_slice(sl);
-            Ok(Midstate(ret))
+    pub fn new(state: [u8; 32], bytes_hashed: usize) -> Result<Self, MidstateError> {
+        if bytes_hashed % 64 != 0 {
+            return Err(MidstateError { invalid_length: bytes_hashed });
         }
+
+        Ok(Midstate {
+            bytes: state,
+            length: bytes_hashed,
+        })
     }
 
-    /// Unwraps the [`Midstate`] and returns the underlying byte array.
-    pub const fn to_byte_array(self) -> [u8; 32] { self.0 }
+    /// Construct a new [`Midstate`] from the `state` and the `bytes_hashed` to get to that state.
+    ///
+    ///
+    /// # Panics
+    ///
+    /// Panics if `bytes_hashed` is not a multiple of 64.
+    pub const fn new_unchecked(state: [u8; 32], bytes_hashed: usize) -> Self {
```

Comment:
It is checked though!

Reply:
In this case I think it'd be fine to call this just `new`. If users are calling this with a non-multiple-of-64 length that is a programming error. Typically the length should even be a known fixed constant. Typically 64.

We could also add a `try_new` variant which returns an `Option` instead of panicking, though I'm unsure how useful it would be.

Reply:
As an aside, we should have an accessor for the length. Maybe `fn n_bytes_hashed` or `n_bits_hashed` or something. Because AFAICT currently users can't even tell how many bytes they've hashed, and we don't want to encourage users to construct whole midstates just to get this information.

Reply:
> If users are calling this with a non-multiple-of-64 length that is a programming error.

Yes, I think so.

> currently users can't even tell how many bytes they've hashed,

They can count manually externally but it's arguably annoying and wasteful, so concept ACK.

Reply:
We have `HashEngine::n_bytes_hashed()` and also, with this PR applied, `HashEngine::can_extract_midstate()` (now `const` thanks to review).